File: app/src/main/java/com/bearcave/passageplanning/passages/PassageManagerAdapter.java
package com.bearcave.passageplanning.passages;

import android.content.Context;
import android.util.SparseArray;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bearcave.passageplanning.R;
import com.bearcave.passageplanning.base.BaseManagerAdapter;
import com.bearcave.passageplanning.base.BaseManagerFragment;
import com.bearcave.passageplanning.data.database.tables.passage.PassageDAO;
import com.bearcave.passageplanning.data.database.tables.waypoints.ReadWaypoints;
import com.bearcave.passageplanning.data.database.tables.waypoints.WaypointDAO;

import java.util.ArrayList;

import butterknife.ButterKnife;


public class PassageManagerAdapter extends BaseManagerAdapter {


    private ReadWaypoints waypointsDatabase;
    private SparseArray<WaypointDAO> waypoints;

    public PassageManagerAdapter(BaseManagerFragment parent, Context context) {
        super(parent, context);
        this.waypoints = new SparseArray<>();
    }


    private WaypointDAO getWaypointById(Integer id, int group){
        WaypointDAO requestedWaypoint = waypoints.get(id);

        if (requestedWaypoint == null){
            ArrayList<WaypointDAO> loaded = (ArrayList<WaypointDAO>) waypointsDatabase.read(getPassages().get(group).getWaypointsIds());
            for (WaypointDAO waypoint: loaded) {
                waypoints.put(id, waypoint);
            }
            return waypoints.get(id);
        }

        return requestedWaypoint;
    }

    private WaypointDAO getWaypointFromList(int group, int child){
        return getWaypointById(getPassages().get(group).getWaypointsIds().get(child), group);
    }

    private ArrayList<PassageDAO> getPassages(){
        return getContainer();
    }


    @Override
    public int getChildrenCount(int groupPosition) {
        return getPassages().get(groupPosition).getWaypointsIds().size();
    }

    @Override
    public Object getChild(int groupPosition, int childPosition) {
        return getWaypointFromList(groupPosition, childPosition);
    }


    @Override
    public View getChildView(int groupPosition, int childPosition, boolean isLastChild, View convertView, ViewGroup parent) {
        View view = getInflater().inflate(R.layout.passage_manager_child_item, parent, false);
        WaypointDAO waypoint = getWaypointFromList(groupPosition, childPosition);

        TextView title = ButterKnife.findById(view, R.id.name);
        title.setText(getPassages().get(groupPosition).getName());

        ImageView up = ButterKnife.findById(view, R.id.up);
        ImageView down = ButterKnife.findById(view, R.id.down);

        return view;
    }



}

File: app/src/main/java/com/bearcave/passageplanning/passages/PassageManagerAdapter.kt
package com.bearcave.passageplanning.passages

import android.content.Context
import android.support.v7.app.AlertDialog
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import butterknife.ButterKnife
import com.bearcave.passageplanning.R
import com.bearcave.passageplanning.base.BaseManagerAdapterWithWaypoints
import com.bearcave.passageplanning.passages.database.Passage

/**
 *
 * @author Michał Wąsowicz
 * @since 27.05.17
 * @version 1.0
 */
class PassageManagerAdapter(parent: PassageManagerFragment, context: Context) : BaseManagerAdapterWithWaypoints<Passage, Int>(parent, context) {

    init {
        addOption("Start passage", { dao ->
            parent.startPassage(dao)
        })

        addOption(context.getString(R.string.action_edit), { dao ->
            parent.openEditor(dao)
        })

        addOption(context.getString(R.string.show_more), { dao ->
            val alertDialog = AlertDialog.Builder(context).create()
            alertDialog.setMessage("""
                |Draught: ${dao.draught}${context.getString(R.string.depth_unit)}
                |Speed: ${dao.speed}${context.getString(R.string.speed_unit)}
            """.trimMargin())
            alertDialog.show()
        })

        addOption(context.getString(R.string.action_delete), { dao ->
            val alertDialog = AlertDialog.Builder(context).create()
            alertDialog.setTitle("Are you sure you want to delete this?")
            alertDialog.setButton(
                    AlertDialog.BUTTON_POSITIVE,
                    "Yes",
                    {
                        _, _ ->
                        run {
                            parent.delete(dao)
                            container.remove(dao)
                            notifyDataSetChanged()
                        }
                    }
            )


            alertDialog.setButton(
                    AlertDialog.BUTTON_NEGATIVE,
                    "No",
                    { _, _ ->  }
            )

            alertDialog.show()
        })
    }

    override fun getChildView(groupPosition: Int, childPosition: Int, isLastChild: Boolean, convertView: View?, parent: ViewGroup?): View {
        val view = convertView ?: inflater.inflate(R.layout.passage_child_item, parent, false)

        val title = ButterKnife.findById<TextView>(view, R.id.waypoint)
        title.text = getWaypointById(
                container[groupPosition]
                        .route
                        .waypointsIds[childPosition]
        ).name

        return view
    }

    override fun getChildrenCount(groupPosition: Int): Int = container[groupPosition].route.waypointsIds.size

}